File: crates/vleo-core/src/physics/cost.rs
//! Cost.
//!
//! Cost estimating relationships are fitted to historical programmes, and every
//! one of them is a power law in a mass or a power with a large residual. They
//! are here because a design tool that cannot say what a decision costs is a
//! tool that will be overruled by a spreadsheet — and the spreadsheet will not
//! have the design in it.
//!
//! The honest statement about accuracy: these are order-of-magnitude at
//! concept stage. They belong at the management layer, downstream of a selected
//! design, and they carry their own credibility like every other node.

use vleo_units::pmath;
use vleo_units::*;

/// A cost estimating relationship of the form `cost = a·x^b`, in a stated
/// currency year.
#[derive(Clone, Copy, Debug)]
pub struct Cer {
    pub a: f64,
    pub b: f64,
    /// The currency year the fit is in. A CER quoted without one is unusable,
    /// and quoting it without one is how a 2005 fit becomes a 2026 estimate.
    pub base_year: u16,
    /// One-sigma residual of the fit, as a fraction. Reported with every
    /// estimate, because a point estimate from a power law with a 40% residual
    /// is not a number, it is a centre.
    pub sigma: f64,
}

impl Cer {
    /// Evaluate, in the CER's own base year currency.
    pub fn evaluate(&self, driver: f64) -> Money {
        Money::new(self.a * pmath::powf(driver, self.b) * 1.0e6)
    }
    /// Evaluate and inflate to a target year at a stated annual rate.
    pub fn evaluate_inflated(&self, driver: f64, target_year: u16, annual_inflation: f64) -> Money {
        let years = (target_year as f64) - (self.base_year as f64);
        Money::new(self.evaluate(driver).get() * pmath::powf(1.0 + annual_inflation, years))
    }
}

/// Wright's learning curve: the `n`-th unit costs `T1·n^(log2(b))`.
///
/// `b` is the learning-curve slope — 0.95 means each doubling of cumulative
/// production costs 95% of the previous doubling. For a constellation this is
/// the term that decides whether forty satellites cost forty times one, and it
/// is the term most concept estimates leave out.
pub fn unit_cost(first_unit: Money, unit_number: f64, slope: f64) -> Money {
    Money::new(first_unit.get() * pmath::powf(unit_number, pmath::log2(slope)))
}

/// Total cost of a production run under a learning curve, summed rather than
/// integrated — the run is tens of units, not thousands, and the integral
/// approximation is wrong by a few per cent at that size.
pub fn production_run_cost(first_unit: Money, units: u32, slope: f64) -> Money {
    let mut total = 0.0;
    for n in 1..=units {
        total += unit_cost(first_unit, n as f64, slope).get();
    }
    Money::new(total)
}

/// Programme cost: non-recurring, plus the production run, plus operations for
/// the mission duration.
pub fn programme_cost(
    non_recurring: Money,
    production: Money,
    annual_operations: Money,
    years: f64,
) -> Money {
    Money::new(non_recurring.get() + production.get() + annual_operations.get() * years)
}

/// Cost per unit of delivered service — the number the management layer
/// actually compares options on.
pub fn cost_per_service_unit(programme_cost: Money, service_units: f64) -> Money {
    Money::new(programme_cost.get() / service_units)
}

/// The saving an air-breathing design buys, against a stored-propellant design
/// that has to be replaced when its propellant runs out.
///
/// This is the commercial argument for the whole programme, so it is a node
/// with a source and evidence like any other rather than a slide.
pub fn replacement_avoided(
    satellite_unit_cost: Money,
    launch_cost: Money,
    replacements_avoided: f64,
) -> Money {
    Money::new((satellite_unit_cost.get() + launch_cost.get()) * replacements_avoided)
}

/// A published spacecraft bus CER, dry mass driven.
///
/// Small-satellite bus recurring cost against dry mass, fitted over the small
/// end of the historical record. Stated with its residual because the residual
/// is larger than most of the design decisions it would be used to compare.
pub const BUS_RECURRING_CER: Cer = Cer { a: 0.98, b: 0.72, base_year: 2020, sigma: 0.40 };

/// Payload recurring cost against payload mass.
pub const PAYLOAD_RECURRING_CER: Cer = Cer { a: 1.85, b: 0.65, base_year: 2020, sigma: 0.45 };

/// Non-recurring engineering against total dry mass.
pub const NON_RECURRING_CER: Cer = Cer { a: 14.2, b: 0.55, base_year: 2020, sigma: 0.55 };

File: tools/nodes/cost.py
#!/usr/bin/env python3
"""Cost.

One subsystem's rows, and nothing else. Split out of the seeder so that
changing what this subsystem declares does not mean opening the file that
declares every other one — the same reason each subsystem has its own crate.

The helpers come from `seed_helpers`, which owns the shared lists these
append to. Import order is authoring order, so `seed_tree` decides it and this
file only says what the rows are.
"""

from seed_helpers import *          # noqa: F401,F403  the row helpers: D, C, layer, S, unnamed

# =============================================================================
# COST
# =============================================================================
D("cost_learning_slope", "Production learning curve slope", "cost", "mgt_cost", "Ratio", "One", "b_lc",
  0.92, 0.7, 1.0,
  "below 0.70 no space production programme has learned that fast",
  "unity is no learning at all, the pessimistic limit",
  "By what factor does unit cost fall with each doubling of cumulative production?",
  "nasa_cer", "programme", "A. Rai / 2026-09-01",
  note="The term that decides whether forty satellites cost forty times one, and the term most concept estimates leave out.")

D("cost_launch_per_satellite", "Launch cost per satellite", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_lch",
  1.8, 0.05, 100.0,
  "below 50 thousand dollars no launch service exists",
  "above 100 million per satellite the constellation is not the one costed here",
  "What does it cost to put one satellite in orbit?",
  "orbitt_case_c1", "programme", "A. Rai / 2026-09-01")

D("cost_annual_operations", "Annual operations cost", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_ops",
  4.5, 0.1, 200.0,
  "below 100 thousand a year no ground segment and team exist",
  "above 200 million a year the programme is not the one costed here",
  "What does running the constellation cost each year?",
  "orbitt_case_c1", "programme", "A. Rai / 2026-09-01")

D("cost_inflation", "Annual inflation rate", "cost", "mgt_cost", "Ratio", "One", "i_r",
  0.030, 0.0, 0.15,
  "a negative rate is deflation, which the cost model has no basis for",
  "above 15% a year the estimate is dominated by the escalation rather than the design",
  "At what annual rate are the fitted cost relations escalated to the target year?",
  "nasa_cer", "programme", "A. Rai / 2026-09-01")

C("cost_bus_recurring", "Bus recurring cost, first unit", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_bus",
  "What does the first spacecraft bus cost to build?",
  "C = a*m_dry^b, escalated",
  "nasa_cer", "programme",
  [("m", "mass_dry"), ("i", "cost_inflation")],
  [("evaluate the bus cost estimating relation on dry mass and escalate to 2026",
    "c", "Money", "cost::BUS_RECURRING_CER.evaluate_inflated(m.get(), 2026, i.get())")],
  0.0, 10000.0,
  "a cost cannot be negative",
  "above 10 billion for one bus the relation is being fed a mass it was not fitted on",
  tier="C",
  assumptions=[("The historical fit applies to this class of spacecraft",
                "the fit's one-sigma residual is 40%, which is larger than most of the design decisions it would be used to compare")])

C("cost_payload_recurring", "Payload recurring cost, first unit", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_pay",
  "What does the first payload set cost to build?",
  "C = a*m_pay^b, escalated",
  "nasa_cer", "programme",
  [("m", "mass_payload"), ("i", "cost_inflation")],
  [("evaluate the payload cost estimating relation on payload mass and escalate to 2026",
    "c", "Money", "cost::PAYLOAD_RECURRING_CER.evaluate_inflated(m.get(), 2026, i.get())")],
  0.0, 10000.0,
  "a cost cannot be negative",
  "above 10 billion for one payload set the relation is being fed a mass it was not fitted on",
  tier="C")

C("cost_non_recurring", "Non-recurring engineering cost", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_nre",
  "What does developing the design cost, once?",
  "C = a*m_dry^b, escalated",
  "nasa_cer", "programme",
  [("m", "mass_dry"), ("i", "cost_inflation")],
  [("evaluate the non-recurring cost estimating relation on dry mass and escalate to 2026",
    "c", "Money", "cost::NON_RECURRING_CER.evaluate_inflated(m.get(), 2026, i.get())")],
  0.0, 100000.0,
  "a cost cannot be negative",
  "above 100 billion the relation is being fed a mass it was not fitted on",
  tier="C")

C("cost_production", "Production run cost", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_prod",
  "What does building the whole constellation cost?",
  "C = sum over n of T1*n^log2(b)",
  "nasa_cer", "programme",
  [("cb", "cost_bus_recurring"), ("cp", "cost_payload_recurring"), ("n", "mis_satellites"),
   ("b", "cost_learning_slope")],
  [("sum the learning curve over the production run of buses and payloads together",
    "c", "Money",
    "cost::production_run_cost(Money::new(cb.get() + cp.get()), n.get() as u32, b.get())")],
  0.0, 1000000.0,
  "a cost cannot be negative",
  "above a trillion dollars the programme is not the one being designed",
  tier="C")

C("cost_programme", "Total programme cost", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_tot",
  "What does the whole programme cost over the mission?",
  "C = NRE + production + launch + operations*years",
  "nasa_cer", "programme",
  [("nre", "cost_non_recurring"), ("pr", "cost_production"), ("lc", "cost_launch_per_satellite"),
   ("n", "mis_satellites"), ("op", "cost_annual_operations"), ("y", "orbit_mission_duration")],
  [("add the non-recurring cost, the production run, the launch of every satellite and the operations over the mission",
    "c", "Money",
    "cost::programme_cost(nre, Money::new(pr.get() + lc.get() * n.get()), op, y.get() / 31_557_600.0)")],
  0.0, 1000000.0,
  "a cost cannot be negative",
  "above a trillion dollars the programme is not the one being designed",
  tier="C",
  kpis=["kpi_cost_per_year"])

C("cost_per_year", "Cost per operational year", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_yr",
  "What does a year of service cost?",
  "C = C_total/years",
  "nasa_cer", "programme",
  [("c", "cost_programme"), ("y", "orbit_mission_duration")],
  [("divide the programme cost by the mission duration in years",
    "cy", "Money",
    "cost::cost_per_service_unit(c, y.get() / 31_557_600.0)")],
  0.0, 1000000.0,
  "a cost cannot be negative",
  "above a trillion dollars a year the programme is not the one being designed",
  tier="C",
  kpis=["kpi_cost_per_year"])

C("cost_replacement_avoided", "Replacement cost avoided", "cost", "mgt_cost", "Money", "MillionUsDollar", "C_avd",
  "What does air-breathing propulsion save by removing the need to replace satellites when propellant runs out?",
  "C = (C_unit + C_launch)*N_replacements",
  "orbitt_case_c1", "programme",
  [("cb", "cost_bus_recurring"), ("cp", "cost_payload_recurring"), ("lc", "cost_launch_per_satellite"),
   ("n", "mis_satellites"), ("td", "prop_thrust_to_drag")],
  [("count one avoided replacement of the whole constellation when thrust exceeds drag, and none when it does not",
    "c", "Money",
    "cost::replacement_avoided(Money::new(cb.get() + cp.get()), lc, if td.get() >= 1.0 { n.get() } else { 0.0 })")],
  0.0, 1000000.0,
  "a saving cannot be negative",
  "above a trillion dollars the programme is not the one being designed",
  tier="C",
  note="The commercial argument for the whole programme, as a node with a source and evidence rather than a slide.")
